Tfc kubernetes integration (#1431)
hashicorp/terraform-website: content/source/docs/cloud/integrations/kubernetes/index.html.md

---
layout: "cloud"
page_title: "Terraform Cloud Operator for Kubernetes setup instructions"
description: |-
  HashiCorp Terraform Cloud customers can integrate with Kubernetes using the official Terraform Cloud Operator for Kubernetes to provision infrastructure through Terraform Cloud using their Kubernetes control plane
---

# Terraform Cloud Operator for Kubernetes Setup Instructions

> -> **Note:** The Terraform Cloud Operator for Kubernetes is still under development and in the alpha testing stage. It is not ready for production use.

Reviewer: ```suggestion
~> **Important:** The Terraform Cloud Operator for Kubernetes is still under development and in the alpha testing stage. It is not ready for production use.
```

I reckon that's worth a yellow box instead of just a blue one.

## Overview

> HashiCorp Terraform Cloud customers can integrate with Kubernetes using the official [Terraform Cloud Operator for Kubernetes](https://github.com/hashicorp/terraform-k8s) to provision infrastructure internal or external to the Kubernetes cluster directly from the Kubernetes control plane.  Using the Terraform Cloud Operator for Kubernetes' CustomResourceDefition (CRD), users can dynamically create Terraform Cloud workspaces, populate variables, and apply a Terraform configuration stored in a git repository or directly from the [Terraform Registry](https://registry.terraform.io/).

Reviewer: - Last sentence was unclear; it probably meant you can use a registry module as the root module for a workspace, so I changed it to keep that associated with the right bit.
- Typo (defition).

```suggestion
HashiCorp Terraform Cloud customers can integrate with Kubernetes using the official [Terraform Cloud Operator for Kubernetes](https://github.com/hashicorp/terraform-k8s) to provision infrastructure internal or external to the Kubernetes cluster directly from the Kubernetes control plane.  Using the Terraform Cloud Operator for Kubernetes' CustomResourceDefinition (CRD), users can dynamically create Terraform Cloud workspaces (using a Terraform configuration from a git repository or from the [Terraform Registry](https://registry.terraform.io/), populate variables, and perform Terraform runs to provision infrastructure.
```

## Prerequisites

> Access and support for the Terraform Cloud Operator for Kubernetes is to all customers in the Terraform Cloud.  Some features of Terraform Cloud that are [limited to certain tiers](/docs/cloud/paid.html) and aren't available to the Terraform Cloud Operator for Kubernetes unless you've purchased the corresponding tier.

Reviewer: ```suggestion
All Terraform Cloud users can use the Terraform Cloud Operator for Kubernetes. Some features of Terraform Cloud that are [limited to certain tiers](/docs/cloud/paid.html) aren't available to the Terraform Cloud Operator for Kubernetes unless you've purchased the corresponding tier.
```

## Terraform Cloud Operator for Kubernetes

### Networking Requirements

In order for the Terraform Cloud Operator for Kubernetes to function properly, it must be able to make outbound requests over HTTPS (TCP port 443) to the Terraform Cloud application APIs. This may require perimeter networking as well as container host networking changes, depending on your environment. The IP ranges are documented in the [Terraform Cloud IP Ranges documentation](/docs/cloud/architectural-details/ip-ranges.html). The services which run on these IP ranges are described in the table below.

Hostname | Port/Protocol | Directionality | Purpose
 --|--|--|--
 app.terraform.io | tcp/443, HTTPS | Outbound | Dynamically managing Terraform Cloud Workspaces and returning the output to Kubernetes via the Terraform Cloud API

### Compatibility

The current release of the Terraform Cloud Operator for Kubernetes supports the following versions:

* Kubernetes 1.14 and above

### Installation & Configuration

> * Generate an [Organization token](/docs/cloud/users-teams-organizations/api-tokens.html#organization-api-tokens) within Terraform Cloud and save it to a `credentials` file

Reviewer: - Since these list points contain complete sentences, they should be punctuated as complete sentences. 
- Since they're ordered steps, they should be a numbered list instead of bulleted. 
- Markup issues: the code  blocks should have empty lines on either side of them, and they should be indented (exactly) four spaces to keep them associated with the list item they belong to. (That results in a code block _inside_ a list item.)
- We can save the reader some effort by making it clear that a "credentials file" doesn't refer to any outside concept they need to be aware of. 

I'll adjust this first item or two, and you can adjust the rest to match. 

```suggestion
1. Generate an [organization token](/docs/cloud/users-teams-organizations/api-tokens.html#organization-api-tokens) within Terraform Cloud and save it to a file. (These instructions assume you're using a file named `credentials`.)
```

> * Create a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/) with the Terraform Cloud API credentials
> ```
> kubectl -n $NAMESPACE create secret generic terraformrc --from-file=credentials
> ```

Reviewer: (numbered lists are easier to maintain if you just use "1" for all the numbers.)

```suggestion

1. Create a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/) with the Terraform Cloud API credentials

    ```
    kubectl -n $NAMESPACE create secret generic terraformrc --from-file=credentials
    ```

```

* Add sensitive variables, such as your cloud provider credentials, to the workspace
```
kubectl -n $NAMESPACE create secret generic workspacesecrets --from-literal=secret_key=abc123
```
* Install the [Terraform Cloud Operator for Kubernetes via Helm](https://github.com/hashicorp/terraform-helm)
```
helm repo add hashicorp https://helm.releases.hashicorp.com
```

> ```

Reviewer: Just use one code block, and put a blank line between the two commands.

helm install --devel --namespace ${RELEASE_NAMESPACE} hashicorp/terraform --generate-name
```
* To create a Terraform workspace, you can create a separate Helm chart to deploy the custom resource or examine this [example](https://github.com/hashicorp/terraform-helm/tree/master/example).

### Upgrading

> The Helm chart automatically installs all Custom Resource Definitions under the `crds/` directory. As a result, any updates to the schema must be manually copied into the directory and removed from the Kubernetes cluster.

Reviewer: Does stuff in Kube-land consistently use the run-together form of "CustomResourceDefinitions"? If so, let's follow suit.

Author: It does.  I'll change on the next update.

```

> kubectl delete crd workspaces.app.terraform.io

Reviewer: This doesn't seem like complete upgrading instructions. What am I missing here?

Author: It was the chart developer instructions.  I changed it to the end user instructions.

## Prerequisites

All Terraform Cloud users can use the Terraform Cloud Operator for Kubernetes. Some features of Terraform Cloud that are [limited to certain tiers](/docs/cloud/paid.html) aren't available to the Terraform Cloud Operator for Kubernetes unless you've purchased the corresponding tier.

## Terraform Cloud Operator for Kubernetes

### Networking Requirements

In order for the Terraform Cloud Operator for Kubernetes to function properly, it must be able to make outbound requests over HTTPS (TCP port 443) to the Terraform Cloud application APIs. This may require perimeter networking as well as container host networking changes, depending on your environment. The IP ranges are documented in the [Terraform Cloud IP Ranges documentation](/docs/cloud/architectural-details/ip-ranges.html). The services which run on these IP ranges are described in the table below.

Hostname | Port/Protocol | Directionality | Purpose
 --|--|--|--
 app.terraform.io | tcp/443, HTTPS | Outbound | Dynamically managing Terraform Cloud Workspaces and returning the output to Kubernetes via the Terraform Cloud API

> ### Compatibility

Reviewer 2: Should we specify helm 3?

The current release of the Terraform Cloud Operator for Kubernetes supports the following versions:

* Kubernetes 1.14 and above

### Installation & Configuration

1. Generate an [organization token](/docs/cloud/users-teams-organizations/api-tokens.html#organization-api-tokens) within Terraform Cloud and save it to a file. (These instructions assume you're using a file named `credentials`.)

> 1. Create a [Kubernetes Secret](https://kubernetes.io/docs/concepts/configuration/secret/) with the Terraform Cloud API credentials

Reviewer: Still need to add a period at the sentence end, for this list item and the two after it.

Reviewer 2: Should we add text here on what a credentials file looks like? Or perhaps link to https://www.terraform.io/docs/commands/cli-config.html#credentials-1

    ```
    kubectl -n $NAMESPACE create secret generic terraformrc --from-file=credentials
    ```

1. Add sensitive variables, such as your cloud provider credentials, to the workspace

    ```
    kubectl -n $NAMESPACE create secret generic workspacesecrets --from-literal=secret_key=abc123
    ```

1. Install the [Terraform Cloud Operator for Kubernetes via Helm](https://github.com/hashicorp/terraform-helm)

    ```
    helm repo add hashicorp https://helm.releases.hashicorp.com

    helm install --devel --namespace ${RELEASE_NAMESPACE} hashicorp/terraform --generate-name
    ```

> 1. To create a Terraform workspace, you can create a separate Helm chart to deploy the custom resource or examine this [example](https://github.com/hashicorp/terraform-helm/tree/master/example).

Reviewer 2: I'd change "this example" to "these examples" since each workspace yaml in the folder is a different example
